x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/esql_autocomplete/use_esql_fields_options.ts

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { EuiComboBoxOptionOption } from '@elastic/eui';
+import type { ExpressionsStart, Datatable } from '@kbn/expressions-plugin/public';
+
+import { useQuery } from '@tanstack/react-query';
+
+import { useKibana } from '@kbn/kibana-react-plugin/public';
+
+import { getEsqlQueryConfig } from '../../logic/get_esql_query_config';
+import type { FieldType } from '../../logic/esql_validator';
+
+export const esqlToOptions = (
+  data: { error: unknown } | Datatable | undefined,
+  fieldType?: FieldType
+) => {
+  if (data && 'error' in data) {
+    return [];
+  }
+
+  const options = (data?.columns ?? [])
+    .filter(({ meta }) => {
+      // if fieldType absent, we do not filter columns by type
+      if (!fieldType) {
+        return true;
+      }
+      return fieldType === meta.type;
+    })
+    .map(({ id }) => ({ label: id }));
```

Comment:
both `reduce` and `filter` + `map` has the same time complexity of O(n), where n is number of elements in array.
Although, true, reduce will traverse array only once as filter+map twice, but still cunts as the same time complexity.

Sometimes, I prefer filter+map in such cases, because it's slightly more readable to me. Generally, I think, it's a matter of personal preference, unless team specifically agreed to use only one approach for consistency of code base

I found a good article around this case: https://www.davideasaf.com/mapfilter-vs-reduce/, that's basically exactly my thoughts, when I was writing this code 😄 
Let me know, what do you think

Reply:
@marshallmain @vitaliidm Locally, with winlogbeat + auditbeat running, I see 1800+ columns coming back resulting in the data being looped over a total of 3440 times. I know for sure infosec has indices with a lot more fields than that, I think over 15000. I know we got hit bad with a similar issue in browserFields [here](https://github.com/elastic/kibana/pull/127037/files#r820799428). I'd prefer we use reduce for safety to avoid any potential issues. I took the initiative to re-implement this with reduce and the complexity seems minimal and the readability, in my opinion, is still sufficient. Using the below code it ran 1800 times instead of 3440.

```nodejs
(data?.columns ?? [])
    .reduce<Array<{ label: string }>>((acc, { meta, id }) => {
      // if fieldType absent, we do not filter columns by type
      if (!fieldType || fieldType === meta.type) {
        acc.push({ label: id })
      }
      return acc;
    }, []);
```

Also, not seeing jest tests for these hooks, do we still want to merge before these are unit tested?

Not trying to block you over this but this is just my thinking since we have the tendency of getting SDH around index field iteration in other areas of the code. Thanks for bearing with me, just trying to do the right thing for our users.

Reply:
Taking a secondary look through this code, I realized the `options` returned from this function should be wrapped in a `useMemo`.

Reply:
Would it be better to do something like the below? In the grand scheme of things I don't think it'll affect performance much, but avoids an unnecessary filter for the nonexistent-fieldtype scenarios

```
const colData = data.columns ?? [];
const options = fieldType
 ? colData.filter(...).map(....)
 : colData.map(...)?
```

Reply:
Oops, had this window open while I was in a meeting, feel free to follow @dhurley14 's suggestion 😄

Reply:
@dhurley14 , thanks for explaining motivation behind using reduce.
I have changed as you suggested.

But, I don't think it can cause any performance issue
Here is some test snippet: https://gist.github.com/vitaliidm/2fbff971f867ff20f5d355158172506b
With 15k columns, both methods yields results in timerange 1-3ms
Even less difference for 3k.

As for SDHs, wasn't it more related to a fact that browser tried to parse large response from request, rather than additional traverse of array?
Also remember, issues were caused by a fact, that time complexity of fields iteration was O(n^2) in some cases (If I remember correctly find on array was called within forEach method)
That's where large number of fields can cause significant increase of computation time

Anyway, that's very interesting topic, thanks for bringing it to attention. Happy to continue discussion on tech time if needed

> Also, not seeing jest tests for these hooks, do we still want to merge before these are unit tested?

I am focusing more on functional/cypress tests right now, but also will do my best to add unit tests as well. Added few already